File: frontend/types/recipe.ts
export enum RecipeType {
  APPETIZER = 'appetizer',
  MAIN_COURSE = 'main_course',
  SIDE_DISH = 'side_dish',
  SOUP = 'soup',
  SALAD = 'salad',
  DESSERT = 'dessert',
  BEVERAGE = 'beverage',
  SAUCE = 'sauce',
  BREAD = 'bread',
  SNACK = 'snack',
}

export enum Difficulty {
  EASY = 'easy',
  MEDIUM = 'medium',
  HARD = 'hard',
}

export interface Ingredient {
  name: string;
  quantity: string;
  unit: string;
  notes?: string;
}

export interface Instruction {
  step_number: number;
  instruction: string;
  time?: number;
}

export interface Nutrition {
  calories?: number;
  protein?: string;
  carbs?: string;
  fat?: string;
}

export interface RecipeSource {
  type: string;
  url?: string;
  platform?: string;
}

export interface Recipe {
  _id: string;
  user_id: string;
  title: string;
  description?: string;
  recipe_type?: RecipeType;
  cuisine?: string;
  dietary_info: string[];
  prep_time?: number;
  cook_time?: number;
  total_time?: number;
  servings?: number;
  difficulty?: Difficulty;
  ingredients: Ingredient[];
  instructions: Instruction[];
  nutrition?: Nutrition;
  images: string[];
  source?: RecipeSource;
  tags: string[];
  notes?: string;
  is_favorite: boolean;
  created_at: string;
  updated_at: string;
}

export interface RecipeCreate {
  url: string;
  notes?: string;
  tags?: string[];
}

export interface RecipeUpdate extends Partial<Omit<Recipe, '_id' | 'user_id' | 'created_at' | 'updated_at'>> {}

// Recipe type labels for UI
export const RECIPE_TYPE_LABELS: Record<RecipeType, string> = {
  [RecipeType.APPETIZER]: 'Appetizer',
  [RecipeType.MAIN_COURSE]: 'Main Course',
  [RecipeType.SIDE_DISH]: 'Side Dish',
  [RecipeType.SOUP]: 'Soup',
  [RecipeType.SALAD]: 'Salad',
  [RecipeType.DESSERT]: 'Dessert',
  [RecipeType.BEVERAGE]: 'Beverage',
  [RecipeType.SAUCE]: 'Sauce & Dressing',
  [RecipeType.BREAD]: 'Bread & Baked Goods',
  [RecipeType.SNACK]: 'Snack',
};

// Recipe type descriptions
export const RECIPE_TYPE_DESCRIPTIONS: Record<RecipeType, string> = {
  [RecipeType.APPETIZER]: 'Small dishes served before the main course',
  [RecipeType.MAIN_COURSE]: 'Primary dish of a meal, usually substantial',
  [RecipeType.SIDE_DISH]: 'Accompaniments to the main course',
  [RecipeType.SOUP]: 'Liquid-based dishes, hot or cold',
  [RecipeType.SALAD]: 'Cold dishes with mixed vegetables, fruits, or proteins',
  [RecipeType.DESSERT]: 'Sweet dishes typically served after a meal',
  [RecipeType.BEVERAGE]: 'Drinks including smoothies, cocktails, teas',
  [RecipeType.SAUCE]: 'Condiments, dressings, and sauces',
  [RecipeType.BREAD]: 'Baked goods including breads, rolls, and pastries',
  [RecipeType.SNACK]: 'Light foods eaten between meals',
};

File: backend/app/models/recipe.py
from typing import List, Optional
from datetime import datetime
from beanie import Document, Indexed
from pydantic import Field, BaseModel


class Ingredient(BaseModel):
    name: str
    quantity: str
    unit: str
    notes: Optional[str] = None


class Instruction(BaseModel):
    step_number: int
    instruction: str
    time: Optional[int] = None  # Time in minutes


class Nutrition(BaseModel):
    calories: Optional[int] = None
    protein: Optional[str] = None
    carbs: Optional[str] = None
    fat: Optional[str] = None


class RecipeSource(BaseModel):
    type: str  # website, youtube, instagram, manual
    url: Optional[str] = None
    platform: Optional[str] = None


class Recipe(Document):
    """Recipe document model for MongoDB"""
    
    # Basic Information
    user_id: Indexed(str)
    title: str
    description: Optional[str] = None
    recipe_type: Optional[str] = None  # appetizer, main_course, side_dish, soup, salad, dessert, beverage, sauce, bread, snack
    cuisine: Optional[str] = None
    
    # Dietary Information
    dietary_info: List[str] = Field(default_factory=list)
    
    # Time Information
    prep_time: Optional[int] = None  # in minutes
    cook_time: Optional[int] = None  # in minutes
    total_time: Optional[int] = None  # in minutes
    
    # Recipe Details
    servings: Optional[int] = None
    difficulty: Optional[str] = None  # easy, medium, hard
    
    # Ingredients and Instructions
    ingredients: List[Ingredient] = Field(default_factory=list)
    instructions: List[Instruction] = Field(default_factory=list)
    
    # Nutrition
    nutrition: Optional[Nutrition] = None
    
    # Media
    images: List[str] = Field(default_factory=list)
    
    # Source Information
    source: Optional[RecipeSource] = None
    
    # Organization
    tags: List[str] = Field(default_factory=list)
    notes: Optional[str] = None
    is_favorite: bool = False
    
    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "recipes"
        indexes = [
            "user_id",
            "title",
            "recipe_type",
            "cuisine",
            "tags",
            "is_favorite",
            "created_at"
        ]
    
    class Config:
        json_schema_extra = {
            "example": {
                "user_id": "user123",
                "title": "Spaghetti Carbonara",
                "description": "Classic Italian pasta dish",
                "recipe_type": "main_course",
                "cuisine": "Italian",
                "dietary_info": ["gluten-free-optional"],
                "prep_time": 15,
                "cook_time": 20,
                "servings": 4,
                "difficulty": "medium",
                "ingredients": [
                    {
                        "name": "Spaghetti",
                        "quantity": "400",
                        "unit": "grams"
                    }
                ],
                "instructions": [
                    {
                        "step_number": 1,
                        "instruction": "Boil water for pasta",
                        "time": 5
                    }
                ],
                "tags": ["pasta", "italian", "quick"]
            }
        }